### PR TITLE
Pipe support

### PR DESCRIPTION
**Pull request by sandhu**

### CODE REVIEW

File: src/crusta/core.clj [hunk 1]
```diff
@@ -117,7 +123,26 @@
...
+       (pmap (fn [[src dest]]
+               (println (pr-str src))
+               (println (pr-str dest))
+               (with-open [out ^InputStream (stdout-stream src)
+                           in ^OutputStream (stdin-stream dest)]
+                 (io/copy out in))))
+       doall)
+  {:processes processes
+   :stdin (:stdin (first processes))
+   :stdout (:stdout (last processes))})
```
Comment (joel7t) [inline src/crusta/core.clj:142]: Should `stdout-stream` and `stdin-stream` be used here?

I guess this is internal, so it doesn't need to.

Comment (joel7t) [inline src/crusta/core.clj:142]: Is there a convenient way to include `:stderr`?

File: src/crusta/core.clj [hunk 2]
```diff
@@ -117,7 +123,26 @@
...
+(defn pipe
+  "Executes a series of 'processes' piping the stdout from one process into
+  the stdin for the next. The stdin for the first process and the stdout for
+  the last process are exposed. Processes are created with 'exec'"
+  [& processes]
+  (->> processes
+       (partition 2 1)
+       (pmap (fn [[src dest]]
+               (println (pr-str src))
+               (println (pr-str dest))
```
Comment (joel7t) [inline src/crusta/core.clj:135]: Debugging `println`s?

File: src/crusta/core.clj [hunk 3]
```diff
@@ -117,7 +123,26 @@
                          :stdout (st/join "\n" (stdout-seq process))
                          :stderr (st/join "\n" (stderr-seq process))}))))))
 
-;;; Implementation
+(defn pipe
+  "Executes a series of 'processes' piping the stdout from one process into
+  the stdin for the next. The stdin for the first process and the stdout for
+  the last process are exposed. Processes are created with 'exec'"
```
Comment (joel7t) [inline src/crusta/core.clj:129]: What's the failure mode if one of the processes dies?

File: src/crusta/core.clj [hunk 4]
```diff
@@ -35,55 +36,60 @@
...
   "Returns the exit code for 'process'. If the process has not exited,
   the call will block until the process exits. An optional :timeout
   in ms can be supplied to prevent blocking forever. ':timeout' will
   be returned if the timeout expires before the process exits."
-  [^Process process & {:keys [timeout]}]
-  (if-not timeout
-    (.waitFor process)
-    (deref (future (.waitFor process)) timeout :timeout)))
+  [process & {:keys [timeout]}]
+  (let [^Process process (:process process)]
```
Comment (joel7t) [inline src/crusta/core.clj:61]: `process` is being shadowed.  Consider renaming input parameter to `process-map`.

File: src/crusta/core.clj (hunk 1 above)
Comment (sandhu) [reply src/crusta/core.clj:142]: Good catch. Fixed.

Comment (sandhu) [reply src/crusta/core.clj:142]: The intermediate processes in the pipe will have their own `stderr` and `exit-code`. So the client would need to query them to determine where and what the error was.

If they don't care, they can pass `:redirect-stderr` to the `exec` and have it merged into the stdout.

File: src/crusta/core.clj (hunk 2 above)
Comment (sandhu) [reply src/crusta/core.clj:135]: Good catch. Fixed.

File: src/crusta/core.clj (hunk 4 above)
Comment (sandhu) [reply src/crusta/core.clj:61]: It's intentional and I often shadow if the original has no further use in the scope, but am happy to change it if it's confusing.

File: src/crusta/core.clj (hunk 3 above)
Comment (sandhu) [reply src/crusta/core.clj:129]: Process dies and that causes a cascading close on the streams.

File: src/crusta/core.clj (hunk 4 above)
Comment (joel7t) [reply src/crusta/core.clj:61]: As long as it's known that this is the design, then that's fine.

**Review by joel7t — APPROVED** [review]

**Review by ct7t — APPROVED** [review]

**Pull request merged**